sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuBubbleClockManager.scala

```diff
@@ -0,0 +1,205 @@
+/*
+ * Copyright (c) 2025, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.nvidia.spark.rapids
+
+import java.util.concurrent.locks.ReentrantLock
+
+import scala.collection.mutable
+
+/**
+ * GpuBubbleTime, a metric that measures periods when the GPU pipeline contains "bubbles" - gaps
+ * where the GPU is underutilized. Currently, whether the GPU is fully utilized or not is
+ * determined by checking if there are any pending requests for the GpuSemaphore through the
+ * `GpuSemaphore.isDeviceBusy` API.
+ *
+ * GpuBubbleClockManager runs on a daemon thread to track GpuBubbleTime for asynchronous
+ * non-device runners across the Spark executor. The manager maintains wall-clock timers that
+ * measure elapsed time during periods when the GPU is underloaded (not busy).
+ *
+ * The manager periodically polls GPU utility through `GpuSemaphore.isDeviceBusy` and coordinates
+ * timer state changes when GPU status transitions occur. All registered timers start
+ * accumulating time when the GPU becomes underloaded and pause when GPU being saturated.
+ *
+ * The update of registered timers are synchronized with a lock to ensure thread-safe access while
+ * minimizing lock contention. The manager uses 1ms polling intervals to balance responsiveness
+ * with CPU overhead, regarding that 1ms is precise enough for measuring wall time.
+ */
+private class GpuBubbleClockManager extends Runnable {
+
+  /**
+   * Periodical loop to check GPU utility and update clocks accordingly.
+   */
+  override def run(): Unit = {
```

Comment:
I am not a big fan of having a thread updating metrics in this way. The code appears to be okay from a polling standpoint, but why not flip it around and have the GpuSemaphore offer a callback each time isDeviceBusy would have changed. Then GpuBubbleClockManager could combine the two. It knows what state the GpuSemaphore is in and when a clock is added it starts the clock in the proper state. It also knows when a clock is removed and can stop it correctly as well. Not extra threads and not polling.

Reply:
Hi @revans2, I've also considered the idea of having the GpuSemaphore offer a callback each time isDeviceBusy would have changed. However, my concern about increasing the main thread's overhead (under some conditions, there may be over 100 Spark Tasks per Executor, which would mean over 100 callbacks) led me to choose the polling approach instead.

I did a rough estimation. If I am not wrong, the extra overhead on the main thread shall be less than 1ms each time triggering all the timers as callbacks. It seems to be quite acceptable?

Reply:
Hi @revans2, finally I flipped the implementation as the event-driven approach. Please help to take a look when you have time.